src/librustc/dep_graph/graph.rs

```diff
@@ -192,11 +202,23 @@ impl DepGraph {
             let mut stable_hasher = StableHasher::new();
             result.hash_stable(&mut hcx, &mut stable_hasher);
 
+            let current_fingerprint = stable_hasher.finish();
+
             assert!(self.fingerprints
                         .borrow_mut()
-                        .insert(key, stable_hasher.finish())
+                        .insert(key, current_fingerprint)
                         .is_none());
 
+            let prev_fingerprint = data.previous.fingerprint_of(&key);
+
+            let color = if Some(current_fingerprint) == prev_fingerprint {
+                DepNodeColor::Green
+            } else {
+                DepNodeColor::Red
+            };
+
+            assert!(data.colors.borrow_mut().insert(key, color).is_none());
```

Comment:
Nit: Although I am often tempted to write an `assert!` like this, I tend to think that putting crucial side-effect operations inside of an `assert!` is bad-style. Too easy to overlook. I think I'd prefer to see something like:

```rust
let old_value = data.colors.borrow_mut().insert(key, color);
assert!(old_value.is_none(), "key {:?} executed twice", key);
```

Reply:
Will do.